spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -474,15 +476,33 @@ public static boolean isLocalityEnabled(FileIO io, String location, CaseInsensit
     return false;
   }
 
-  public static boolean isVectorizationEnabled(Map<String, String> properties, CaseInsensitiveStringMap readOptions) {
-    String batchReadsSessionConf = SparkSession.active().conf()
-        .get("spark.sql.iceberg.vectorization.enabled", null);
-    if (batchReadsSessionConf != null) {
-      return Boolean.valueOf(batchReadsSessionConf);
+  public static boolean isVectorizationEnabled(FileFormat fileFormat,
+                                               Map<String, String> properties,
+                                               RuntimeConfig sessionConf,
+                                               CaseInsensitiveStringMap readOptions) {
+
+    String readOptionValue = readOptions.get(SparkReadOptions.VECTORIZATION_ENABLED);
+    if (readOptionValue != null) {
+      return Boolean.parseBoolean(readOptionValue);
```

Comment:
Looks like this reverses the order of precedence. Before, `spark.sql.iceberg.vectorization.enabled` could be used to disable all vectorized reads, even if the read option was set to specifically enable them. Now, the read option will override.

I'm not sure what the right behavior is. The read option seems "closer" in that it overrides global defaults. But the read option is likely compiled into job jars, so the one you can change at runtime would be `spark.sql.iceberg.vectorization.enabled`. I would lean toward keeping the existing behavior instead of changing it, but I could be convinced otherwise.

Reply:
I see it a bit differently, actually. I think we break the config hierarchy right now. If someone made an effort and set a read option (which is the most low-level config), don't we have to respect that? I'd interpret such configs like no matter what my SQL conf is, I know I want to use this value in this particular job.

Like in my [example](https://github.com/apache/iceberg/pull/2248#discussion_r580612852) above, the current order of precedence prohibits using SQL conf for common defaults. For example, Iceberg is used by multiple orgs and each org frequently has its own SQL conf template. Let's say somebody accesses multiple tables in one Spark app and wants to use the default config for all tables except one. Instead of adding a specific read/write option for a single job, he/she would have to modify every single job in the Spark app.

I think users that want to turn on and off certain feature using SQL, should not set read/write options. Instead, rely solely on the SQL conf.

Reply:
I wonder if it is actually only my interpretation. Do you guys see it differently, @rdblue @shardulm94 @RussellSpitzer?

Reply:
My order like I said above is based on a "notebook/cli" perspective, It's also how we set up things in the Spark Cassandra connector.

Did you have to set it for this particular read? Highest Precedent
Did you have to set it for this particular session? Second Highest
Did you set it this in the underlying table config or in the Spark Conf? Last

(More accurately we had a more complex hierarchy because we add session and session scoped parameters which i think was a bit overkill.)

Reply:
I think it should be safe to do this because it is unlikely that a job actually sets this in its config unless it is testing or already tested. So while I think this prohibits the use case of finding a problem and disabling vectorized reads everywhere to be safe, it is unlikely that this should make it into final compiled artifacts. So let's just go with what is here since it appears that is the more common way of thinking about it.

Reply:
FYI @samarthjain